src/memkind_memtier.c

```diff
@@ -881,7 +931,14 @@ MEMKIND_EXPORT void memtier_kind_free(memkind_t kind, void *ptr)
     if (memtier_kind_free_pre)
         memtier_kind_free_pre(&ptr);
 #endif
+
     if (!kind) {
+        // TODO: checking policicy this way is a really ugly hack
+        if (g_mtt_allocator) {
```

Comment:
Two issues:

1. Does destruction of the kind reset the pointer to NULL ? If not, we would call mtt_allocator_free with invalid mtt_allocator - it's a simplification, but I want to point to an issue ...
2. What if multiple kinds are created? Would it be possible to free the pointer with a wrong allocator? Maybe we should call memkind_detect_kind first? It might be a good idea to explain this issue